libgalois/src/analytics/matrix_completion/matrix_completion.cpp

```diff
@@ -0,0 +1,426 @@
+        int round, MatrixCompletionPlan plan) const = 0;
+    virtual std::string Name() const = 0;
+    virtual bool IsBold() const { return false; }
+    virtual ~StepFunction() {}
+  };
+
+  struct PurdueStepFunction : public StepFunction {
+    virtual std::string Name() const { return "Purdue"; }
+    virtual LatentValue StepSize(int round, MatrixCompletionPlan plan) const {
+      return plan.learningRate() * 1.5 /
+             (1.0 + plan.decayRate() * pow(round + 1, 1.5));
+    }
+  };
+
+  struct IntelStepFunction : public StepFunction {
+    virtual std::string Name() const { return "Intel"; }
+    virtual LatentValue StepSize(int round, MatrixCompletionPlan plan) const {
+      return plan.learningRate() * pow(plan.decayRate(), round);
+    }
+  };
+
+  struct BottouStepFunction : public StepFunction {
+    virtual std::string Name() const { return "Bottou"; }
+    virtual LatentValue StepSize(int round, MatrixCompletionPlan plan) const {
+      return plan.learningRate() /
+             (1.0 + plan.learningRate() * plan.lambda() * round);
+    }
+  };
+
+  struct InverseStepFunction : public StepFunction {
+    virtual std::string Name() const { return "Inverse"; }
+    virtual LatentValue StepSize(int round, MatrixCompletionPlan) const {
+      return 1.0 / (round + 1);
+    }
+  };
+
+  struct BoldStepFunction : public StepFunction {
+    virtual std::string Name() const { return "Bold"; }
+    virtual bool IsBold() const { return true; }
+    virtual LatentValue StepSize(int, MatrixCompletionPlan) const {
+      return 0.0;
+    }
+  };
+
+  katana::Result<StepFunction*> NewStepFunction(MatrixCompletionPlan plan) {
+    switch (plan.learningRateFunction()) {
+    case MatrixCompletionPlan::kIntel:
+      return new IntelStepFunction;
+    case MatrixCompletionPlan::kPurdue:
+      return new PurdueStepFunction;
+    case MatrixCompletionPlan::kBottou:
+      return new BottouStepFunction;
+    case MatrixCompletionPlan::kInverse:
+      return new InverseStepFunction;
+    case MatrixCompletionPlan::kBold:
+      return new BoldStepFunction;
+    default:
+      return KATANA_ERROR(
+          katana::ErrorCode::InvalidArgument, "Unknown step function");
+    }
+  }
+
+  double CountFlops(size_t nnz, int rounds, int k, MatrixCompletionPlan plan) {
+    double flop = 0;
+    if (plan.useExactError()) {
+      // dotProduct = 2K, square = 1, sum = 1
+      flop += nnz * (2.0 * k + 1 + 1);
+    } else {
+      // Computed during gradient update: square = 1, sum = 1
+      flop += nnz * (1 + 1);
+    }
+    // dotProduct = 2K, gradient = 10K,
+    flop += rounds * (nnz * (12.0 * k));
+    return flop;
+  }
+
+  size_t InitializeGraphData(Graph& graph, MatrixCompletionPlan plan) {
+    katana::StatTimer initTimer("InitializeGraph");
+    initTimer.start();
+    double top = 1.0 / std::sqrt(LATENT_VECTOR_SIZE);
+    katana::PerThreadStorage<std::mt19937> gen;
+
+#if __cplusplus >= 201103L || defined(HAVE_CXX11_UNIFORM_INT_DISTRIBUTION)
```

Comment:
This says `HAVE_CXX11_UNIFORM_INT_DISTRIBUTION` but then you use `uniform_real_distribution`.

Reply:
This is one of the parts that I have just copied from the previous code. I'm not sure what changes need to be made here. I'll ask @gurbinder533 about it.

Reply:
I wasn't clear in my original message. This code might be correct, but you should check the C++ standard/docs to make sure it actually is. Sorry I wasn't clear.

Reply:
Thanks for the clarification, I asked and since our codebase is C++17 now, that macro is not needed anymore. I fixed it in the last commit.